src/tds/convert.c

```diff
@@ -78,23 +78,23 @@ static const char *parse_numeric(const char *buf, const char *pend,
 
 #define test_alloc(x) {if ((x)==NULL) return TDS_CONVERT_NOMEM;}
 
-#define IS_TINYINT(x) ( 0 <= (x) && (x) <= 0xff )
-#define IS_SMALLINT(x) ( -32768 <= (x) && (x) <= 32767 )
-#define IS_USMALLINT(x) ( 0 <= (x) && (x) <= 65535 )
+#define IS_TINYINT(x) ( 0 <= (x) && (x) < 0x100 )
+#define IS_SMALLINT(x) ( -32768 <= (x) && (x) < 32768 )
+#define IS_USMALLINT(x) ( 0 <= (x) && (x) < 65536 )
 /*
  * f77: I don't write -2147483648, some compiler seem to have some problem 
  * with this constant although is a valid 32bit value
  */
 #define TDS_INT_MIN (-2147483647l-1l)
 #define TDS_INT_MAX 2147483647l
-#define IS_INT(x) (TDS_INT_MIN <= (x) && (x) <= TDS_INT_MAX)
+#define IS_INT(x) (TDS_INT_MIN <= (x) && (x) < (int64_t) TDS_INT_MAX + 1)
 
 #define TDS_UINT_MAX 4294967295lu
-#define IS_UINT(x) (0 <= (x) && (x) <= TDS_UINT_MAX)
+#define IS_UINT(x) (0 <= (x) && (x) < (int64_t) TDS_UINT_MAX + 1)
 
 #define TDS_INT8_MAX ((((TDS_INT8) 0x7fffffffl) << 32) + (TDS_INT8) 0xfffffffflu)
 #define TDS_INT8_MIN (-(((TDS_INT8)1)<<62) -(((TDS_INT8)1)<<62))
-#define IS_INT8(x) (TDS_INT8_MIN <= (x) && (x) <= TDS_INT8_MAX)
+#define IS_INT8(x) (TDS_INT8_MIN <= (x) && (x) < (uint64_t) TDS_INT8_MAX + 1)
```

Comment:
It should be fine... but `IS_INT8(-1)` is false. It's fine in the sense that we should use this to check unsigned numbers and floating point where the check is correct. But if this is used for signed numbers it will fail.

Reply:
Good point.  I suppose an explicit comparison with zero is in order, though doing it up front should at least keep the total number of comparisons at two:
`#define IS_INT8(x) (((x) < 0) ? TDS_INT8_MIN <= (x) : (x) < (uint64_t) TDS_INT8_MAX + 1)`